astropy/coordinates/body.py

```diff
@@ -0,0 +1,560 @@
+        if hasattr(obj, "_flattening"):
+            self._flattening = obj._flattening
+
+    def __len__(self):
+        if self.shape == ():
+            raise IndexError("0-d EarthLocation arrays cannot be indexed")
+        else:
+            return super().__len__()
+
+    def _to_value(self, unit, equivalencies=[]):
+        """Helper method for to and to_value."""
+        # Conversion to another unit in both ``to`` and ``to_value`` goes
+        # via this routine. To make the regular quantity routines work, we
+        # temporarily turn the structured array into a regular one.
+        array_view = self.view(self._array_dtype, np.ndarray)
+        if equivalencies == []:
+            equivalencies = self._equivalencies
+        new_array = self.unit.to(unit, array_view, equivalencies=equivalencies)
+        return new_array.view(self.dtype).reshape(self.shape)
+
+
+geodetic_base_doc = """{__doc__}
+
+    Parameters
+    ----------
+    lon, lat : angle-like
+        The longitude and latitude of the point(s), in angular units. The
+        latitude should be between -90 and 90 degrees, and the longitude will
+        be wrapped to an angle between 0 and 360 degrees. These can also be
+        instances of `~astropy.coordinates.Angle` and either
+        `~astropy.coordinates.Longitude` not `~astropy.coordinates.Latitude`,
+        depending on the parameter.
+    height : `~astropy.units.Quantity` ['length']
+        The height to the point(s).
+    copy : bool, optional
+        If `True` (default), arrays will be copied. If `False`, arrays will
+        be references, though possibly broadcast to ensure matching shapes.
+
+"""
+
+
+@format_doc(geodetic_base_doc)
+class BaseGeodeticRepresentation(BaseRepresentation):
+    """Base geodetic representation."""
+
+    attr_classes = {"lon": Longitude, "lat": Latitude, "height": u.Quantity}
+
+    def __init_subclass__(cls, **kwargs):
+        super().__init_subclass__(**kwargs)
+        if "_ellipsoid" in cls.__dict__:
+            ELLIPSOIDS[cls._ellipsoid] = cls
+        elif (
+            "_equatorial_radius" not in cls.__dict__
+            or "_flattening" not in cls.__dict__
+        ):
+            raise AttributeError(
+                f"{cls.__name__} requires `_ellipsoid` or _equatorial_radius and _flattening."
+            )
+
+    def __init__(self, lon, lat=None, height=None, copy=True):
+        if height is None and not isinstance(lon, self.__class__):
+            height = 0 << u.m
+
+        super().__init__(lon, lat, height, copy=copy)
+        if not self.height.unit.is_equivalent(u.m):
+            raise u.UnitTypeError(
+                f"{self.__class__.__name__} requires height with units of length."
+            )
+
+    def to_cartesian(self):
+        """
+        Converts geodetic coordinates to 3D rectangular (geocentric)
+        cartesian coordinates.
+        """
+        if hasattr(self, "_flattening"):
```

Comment:
If we follow my suggestion above, then `to_cartesian` will only need this branch, since `flattening` and `equatorial_radius` will always be defined.